homeassistant/components/tis_control/switch.py

```diff
@@ -0,0 +1,167 @@
+from TISControlProtocol.mock_api import TISApi
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import MATCH_ALL, STATE_OFF, STATE_ON, Platform
+from homeassistant.core import Event, HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_devices: AddEntitiesCallback
+) -> None:
+    """Set up the TIS switches."""
+    tis_api: TISApi = hass.data[DOMAIN]["tis_api"]
+    # Fetch all switches from the TIS API
+    switches: dict = await tis_api.get_entities(platform=Platform.SWITCH)
+    if switches:
+        # Prepare a list of tuples containing necessary switch details
+        switch_entities = [
+            (
+                appliance_name,
+                next(iter(appliance["channels"][0].values())),
+                appliance["device_id"],
+                appliance["is_protected"],
+                appliance["gateway"],
+            )
+            for switch in switches
+            for appliance_name, appliance in switch.items()
+        ]
+        # Create TISSwitch objects and add them to Home Assistant
+        tis_switches = [
+            TISSwitch(tis_api, switch_name, channel_number, device_id, gateway)
+            for switch_name, channel_number, device_id, is_protected, gateway in switch_entities
+        ]
+        async_add_devices(tis_switches)
+
+
+class TISSwitch(SwitchEntity):
+    """Representation of a TIS switch."""
+
+    def __init__(
+        self,
+        tis_api: TISApi,
+        switch_name: str,
+        channel_number: int,
+        device_id: list[int],
+        gateway: str,
+    ) -> None:
+        """Initialize the switch."""
+        self.api = tis_api
+        self._name = switch_name
+        self._attr_unique_id = f"switch_{self.name}"
+        self.name = switch_name
+        self.device_id = device_id
+        self.gateway = gateway
+        self.channel_number = int(channel_number)
+        self.listener = None
+        self.on_packet = build_packet(
+            operation_code=[0x00, 0x31],
+            ip_address=self.api.host,
+            destination_mac="FF:FF:FF:FF:FF:FF",
+            device_id=self.device_id,
+            additional_packets=[self.channel_number, 0x64, 0x00, 0x00],
+        )
+        self.off_packet = build_packet(
+            operation_code=[0x00, 0x31],
+            ip_address=self.api.host,
+            destination_mac="FF:FF:FF:FF:FF:FF",
+            device_id=self.device_id,
+            additional_packets=[self.channel_number, 0x00, 0x00, 0x00],
+        )
+        self.update_packet = build_packet(
+            operation_code=[0x00, 0x33],
+            ip_address=self.api.host,
+            destination_mac="FF:FF:FF:FF:FF:FF",
+            device_id=self.device_id,
+            additional_packets=[],
+        )
```

Comment:
This is too much protocol info that should be put away in the library imo

Reply:
the problem is that there will be multiple carients of the switch platforms each will have a specific packet so it'd be better  to generate the packet in the platform.
if this would prevent the merge, i could try to figure out a solution but it could take some time